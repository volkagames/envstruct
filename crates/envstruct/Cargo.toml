[package]
name = "envstruct"
resolver = "2"
authors = { workspace = true }
categories = { workspace = true }
description = { workspace = true }
documentation = { workspace = true }
edition = { workspace = true }
keywords = { workspace = true }
license = { workspace = true }
version = { workspace = true }


[dependencies]
bytesize = { version = "1", optional = true }
chrono = { version = "0.4", optional = true }
humantime = { version = "2", optional = true }
regex = { version = "1", optional = true }
thiserror = "2"
url = { version = "2", optional = true }
envstruct_derive = { path = "../envstruct_derive", version = "1.0" }
prettytable-rs = "0.10"
paste = "1.0.15"
serde_json = { version = "1", optional = true }
serde = { version = "1", optional = true }

[dev-dependencies]
strum = { version = "0.26", features = ["derive"] }
serde = { version = "1", features = ["derive"] }
trybuild = "1.0.101"
serial_test = "3.2.0"

[features]
default = [
	"bytesize",
	"chrono",
	"humantime",
	"regex",
	"serde_json",
	"url",
	"env_uppercase",
]
bytesize = ["dep:bytesize"]
chrono = ["dep:chrono"]
humantime = ["dep:humantime"]
regex = ["dep:regex"]
serde_json = ["dep:serde_json", "dep:serde"]
url = ["dep:url"]
env_uppercase = []

[[example]]
name = "usage"
path = "example/usage.rs"

[[example]]
name = "basic"
path = "example/basic.rs"
