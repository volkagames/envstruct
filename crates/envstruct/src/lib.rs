#![allow(internal_features)]

mod env_json;
mod env_map;
mod error;
mod parse_nested;
mod parse_primitive;
mod usage;
mod with_json;

pub use error::*;
pub use parse_nested::*;
pub use parse_primitive::*;
pub use usage::*;

pub use envstruct_derive::*;

/// The `prelude` module re-exports common items for easy inclusion.
pub mod prelude {
    pub use super::{
        env_json::*, env_map::*, error::*, parse_nested::*, parse_primitive::*, usage::*,
        with_json::*,
    };
    pub use envstruct_derive::*;
}

// re-export

/// Re-export of the `bytesize` crate if the `bytesize` feature is enabled.
#[cfg(feature = "bytesize")]
pub use bytesize::{self, ByteSize};

/// Re-export of the `chrono` crate if the `chrono` feature is enabled.
#[cfg(feature = "chrono")]
pub use chrono::{self, DateTime, TimeZone, Utc};

/// Re-export of the `humantime` crate if the `humantime` feature is enabled.
#[cfg(feature = "humantime")]
pub use humantime::{self, Duration};

/// Re-export of the `url` crate if the `url` feature is enabled.
#[cfg(feature = "url")]
pub use url::{self, Url};

/// Re-export of the `regex` crate if the `regex` feature is enabled.
#[cfg(feature = "regex")]
pub use regex::{self, Regex};

/// Re-export of the `serde_json` crate if the `serde_json` feature is enabled.
#[cfg(feature = "serde_json")]
pub use serde_json::{self, Value};
