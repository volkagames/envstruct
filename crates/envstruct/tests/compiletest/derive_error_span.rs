use envstruct::prelude::*;

pub struct Foo {}

#[derive(EnvStruct)]
pub struct Config {
    foo: Foo,
}

fn main() {}
