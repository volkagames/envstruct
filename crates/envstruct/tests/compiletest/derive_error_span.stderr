error[E0599]: no function or associated item named `parse_from_env_var` found for struct `Foo` in the current scope
 --> tests/compiletest/derive_error_span.rs:7:10
  |
3 | pub struct Foo {}
  | -------------- function or associated item `parse_from_env_var` not found for this struct
...
7 |     foo: Foo,
  |          ^^^ function or associated item not found in `Foo`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `parse_from_env_var`, perhaps you need to implement one of them:
          candidate #1: `envstruct::EnvParseNested`
          candidate #2: `envstruct::EnvParsePrimitive`

error[E0599]: no function or associated item named `get_env_entries` found for struct `Foo` in the current scope
 --> tests/compiletest/derive_error_span.rs:7:10
  |
3 | pub struct Foo {}
  | -------------- function or associated item `get_env_entries` not found for this struct
...
7 |     foo: Foo,
  |          ^^^ function or associated item not found in `Foo`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `get_env_entries`, perhaps you need to implement one of them:
          candidate #1: `envstruct::EnvParseNested`
          candidate #2: `envstruct::EnvParsePrimitive`
